library/alloc/src/collections/btree/node.rs

```diff
@@ -1218,16 +1204,16 @@ impl<'a, K: 'a, V: 'a> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Internal>,
         unsafe {
             let mut new_node = Box::new(InternalNode::new());
 
-            let (k, v, new_len) = self.leafy_split(&mut new_node.data);
-            let height = self.node.height;
-            let old_node = &*self.node.as_internal_ptr();
-
+            let (k, v, new_len) = self.transfer_keys_vals_for_split(&mut new_node.data);
             ptr::copy_nonoverlapping(
-                old_node.edges.as_ptr().add(self.idx + 1),
-                new_node.edges.as_mut_ptr(),
+                self.node.edge_at(self.idx + 1),
+                MaybeUninit::slice_as_mut_ptr(&mut new_node.edges),
                 new_len + 1,
             );
+            self.node.as_leaf_mut().len = self.idx as u16;
+            new_node.data.len = new_len as u16;
```

Comment:
I am not sure why this bit was moved out of the transfer function -- it seems like a key part of the transfer to update lengths on both nodes? Is it because of the edge copying just above here?

Reply:
Changing the length intermingled with manipulating the arrays is so confusing. Right here, you can't use `self.node.edge_at(self.idx + 1)` if the edges array is already shortened because it debug_assert's that the index is valid.

Reply:
Here's an alternative with another support function `split_new_node_len`.